shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs

```diff
@@ -36,97 +35,79 @@ public static string GetTypeDisplayName(object item, bool fullName = true)
 
         public static string GetTypeDisplayName(Type type, bool fullName = true)
         {
-            var sb = new StringBuilder();
-            ProcessTypeName(type, sb, fullName);
-            return sb.ToString();
+            var builder = new StringBuilder();
+            ProcessTypeName(builder, type, fullName);
+            return builder.ToString();
         }
 
-        private static void AppendGenericArguments(Type[] args, int startIndex, int numberOfArgsToAppend, StringBuilder sb, bool fullName)
+        private static void ProcessTypeName(StringBuilder builder, Type type, bool fullName)
         {
-            var totalArgs = args.Length;
-            if (totalArgs >= startIndex + numberOfArgsToAppend)
+            if (type.IsGenericType)
             {
-                sb.Append("<");
-                for (int i = startIndex; i < startIndex + numberOfArgsToAppend; i++)
-                {
-                    ProcessTypeName(args[i], sb, fullName);
-                    if (i + 1 < startIndex + numberOfArgsToAppend)
-                    {
-                        sb.Append(", ");
-                    }
-                }
-                sb.Append(">");
+                ProcessGenericType(builder, type, new ArraySegment<Type>(type.GetGenericArguments()), fullName);
+            }
+            else if (_builtInTypeNames.ContainsKey(type))
+            {
+                builder.Append(_builtInTypeNames[type]);
+            }
+            else
+            {
+                builder.Append(fullName ? type.FullName : type.Name);
             }
         }
 
-        private static void ProcessTypeName(Type t, StringBuilder sb, bool fullName)
+        private static void ProcessGenericType(StringBuilder builder, Type type, ArraySegment<Type> genericArguments, bool fullName)
```

Comment:
Let's remove ArraySegment for now. Change return type to int which will be count of how many generic args are printed.
The method will call recursively for nested type. So when we are printing current type, we know offset from recursive calls. We compare offset with current type's GetGenericArguments() to compute how many generic types it has.